Provision GitHub OAuth using secret (#1773)
codenvy/che-docs: modules/end-user-guide/partials/proc_configuring-github-oauth.adoc

OAuth for GitHub allows for automatic SSH key upload to GitHub.

.Prerequisites

> * A running instance of {prod}. To install an instance of {prod}, see {link-installing-an-instance}.

Reviewer: We do not really need a running instance of Che. We need to know Che and Keyckloak URL though.

* The `{orch-cli}` tool is available.

.Procedure

ifeval::["{project-context}" == "che"]

> * For {prod-short} deployed in multi-user mode:

Reviewer: The configuration of the GitHub OAuth application is common to both multi-user and single-user mode. This line should be moved below.

endif::[]

> . Set up the link:https://developer.github.com/apps/building-oauth-apps/creating-an-oauth-app[GitHub OAuth client] with `https://{prod-deployment}-{prod-namespace}.<DOMAIN>/` and `https://keycloak-{prod-namespace}.<DOMAIN>/auth/realms/{prod-deployment}/broker/github/endpoint` as Homepage and Authorization callback urls correspondingly, where `<DOMAIN>` is {orch-name} cluster domain.

Reviewer 2: ```suggestion
. Set up the link:https://developer.github.com/apps/building-oauth-apps/creating-an-oauth-app[GitHub OAuth client] with `https://{prod-deployment}-{prod-namespace}.<DOMAIN>/` and `https://keycloak-{prod-namespace}.<DOMAIN>/auth/realms/{prod-deployment}/broker/github/endpoint` as Homepage and Authorization callback URLs correspondingly, where `<DOMAIN>` is {orch-name} cluster domain.
```

> . Set up the link:https://developer.github.com/apps/building-oauth-apps/creating-an-oauth-app[GitHub OAuth client] with `https://{prod-deployment}-{prod-namespace}.<DOMAIN>/` and `https://keycloak-{prod-namespace}.<DOMAIN>/auth/realms/{prod-deployment}/broker/github/endpoint` as Homepage and Authorization callback URLs correspondingly, where `<DOMAIN>` is {orch-name} cluster domain.

Reviewer: Instead of "Set up the GitHub OAuth client" it would be better to say "Create a new OAuth App in GitHub".

And the URLs you are providing are the default ones (`{prod-deployment}-{prod-namespace}`) but the user may use a different deployment name and namespace.

 ```suggestion
. Create a link:https://developer.github.com/apps/building-oauth-apps/creating-an-oauth-app[OAuth application in GitHub] using {prod-short} URL as the value for the application `Homepage URL` and {identity-provider} GitHub endpoint URL as the value for Authorization callback URL. The default values are `https://{prod-deployment}-{prod-namespace}.<DOMAIN>/` and `https://keycloak-{prod-namespace}.<DOMAIN>/auth/realms/{prod-deployment}/broker/github/endpoint` respectively, where `<DOMAIN>` is {orch-name} cluster domain.
```

. Create a new secret in the {orch-namespace} where {prod-short} is deployed.
+
[subs="+quotes,+attributes"]
----
$ {orch-cli} apply -f - <<EOF
kind: Secret
apiVersion: v1
metadata:
  name: github-oauth-credentials

>   namespace: {prod-namespace}

Reviewer: ```suggestion
  namespace: <...> <1>
```

  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: keycloak-secret
  annotations:
    che.eclipse.org/github-oauth-credentials: 'true'
    che.eclipse.org/mount-as: env
    che.eclipse.org/id_env-name: GITHUB_CLIENT_ID
    che.eclipse.org/secret_env-name: GITHUB_SECRET
data:

>   id: <...> <1>

Reviewer: ```suggestion
  id: <...> <2>
```

>   secret: <...> <2>

Reviewer: ```suggestion
  secret: <...> <3>
```

type: Opaque
EOF
----

> <1> base64 encoded GitHub OAuth Client ID

Reviewer: ```suggestion
<1> {prod-short} namespace. The default is {prod-namespace}
<2> base64 encoded GitHub OAuth Client ID
```

> <2> base64 encoded GitHub OAuth Client Secret

Reviewer: ```suggestion
<3> base64 encoded GitHub OAuth Client Secret
```

> . Wait until rollout of {identity-provider} component finishes.

Reviewer: ```suggestion
. If {prod-short} was already installed wait until rollout of {identity-provider} component finishes.
```
